e2e/nomostest/testing/wrapper.go

```diff
@@ -14,18 +14,70 @@
 
 package testing
 
+import (
+	"strings"
+
+	"kpt.dev/configsync/e2e"
+)
+
 const errorPrefix = "ERROR:"
 
 // Wrapper implements NTB.
 type Wrapper struct {
 	t NTB
 }
 
+// NewShared creates a new instance of the Testing Wrapper that provides additional
+// functionality beyond the standard testing package.
+// It doesn't require the testFeature param.
+func NewShared(t NTB) *Wrapper {
+	t.Helper()
+
+	testFeaturesFlag := *e2e.TestFeatures
+	if testFeaturesFlag != "" {
+		features := strings.Split(testFeaturesFlag, ",")
+		for _, f := range features {
+			sanitizedFeature := strings.TrimSpace(f)
+			if !KnownFeature(Feature(sanitizedFeature)) {
+				t.Fatalf("Test failed because the test-features flag has an unknown feature %q", sanitizedFeature)
+			}
+		}
+	}
+
+	tw := &Wrapper{
+		t: t,
+	}
+	return tw
+}
+
 // New creates a new instance of the Testing Wrapper that provides additional
 // functionality beyond the standard testing package.
-func New(t NTB) *Wrapper {
+func New(t NTB, testFeature Feature) *Wrapper {
 	t.Helper()
 
+	if !KnownFeature(testFeature) {
+		t.Fatalf("Test failed because the feature %q is unknown", testFeature)
+	}
+
+	skip := true
+	testFeaturesFlag := *e2e.TestFeatures
+	if testFeaturesFlag == "" {
+		skip = false
+	} else {
+		features := strings.Split(testFeaturesFlag, ",")
+		for _, f := range features {
+			sanitizedFeature := strings.TrimSpace(f)
+			if !KnownFeature(Feature(sanitizedFeature)) {
+				t.Fatalf("Test failed because the test-features flag has an unknown feature %q", sanitizedFeature)
+			} else if Feature(f) == testFeature {
+				skip = false
+			}
+		}
+	}
```

Comment:
Considering both `New` and `NewShared` checks whether the `TestFeatures` flag includes any unknown features,let us move the logic of checking  the `TestFeatures` flag into `TestMain`. 

We can move the logic of checking whether `testFeature` is supported into the `New` function in`e2e/nomostest/new.go`.

After these two changes, we no longer need both `New` and `NewShared` in this file, can just keep the `func New(t NTB) *Wrapper {` func.

Reply:
> We have to split *e2e.TestFeatures into an array and iterate through the array to search for testFeature. If we move it to TestMain, the logic has to repeat twice.

You can split *e2e.TestFeatures into an array once and track the array somehow. Then the `New` function in `e2e/nomostest/new.go` can reuse the array instead of repeating the logic.

Reply:
That requires us to have a global variable. Even with the global variable, we still need to iterate through the array twice, once in `TestMain` and once in `New`. Why don't we have two separate functions, but have the logic executed only once?

Reply:
To me, the logic in these two functions duplicate a lot, which is why I made the recommendation: https://github.com/GoogleContainerTools/kpt-config-sync/pull/108#discussion_r973513316.

Reply:
I'm not a big fan of using global variable to track an internal value across packages, but like you said, it's a personal preference. If you insist, I'll update this PR.

Reply:
Just notice `New` and `NewShared` cannot be merged into a single function, because `NewShared` doesn't take `testFeature` as an argument, but `New` does.

Reply:
If we go with [comment](https://github.com/GoogleContainerTools/kpt-config-sync/pull/108#discussion_r973513316) , the changes include:
1. Declare a global variable in `e2e/nomostest/testing/feature.go`: `var TestFeatureArray []Feature`. The variable has to be declared here to avoid cyclic import, but it is intialized in `TestMain`, a different package.
2. Check `e2e.TestFeatures` in `TestMain`:
```
	testFeaturesFlag := *e2e.TestFeatures
	if testFeaturesFlag != "" {
		features := strings.Split(testFeaturesFlag, ",")
		for _, f := range features {
			sanitizedFeature := strings.TrimSpace(f)
			nomostesting.TestFeatureArray = append(nomostesting.TestFeatureArray, nomostesting.Feature(sanitizedFeature))
			if !nomostesting.KnownFeature(nomostesting.Feature(sanitizedFeature)) {
				fmt.Printf("Error: Test failed because the test-features flag has an unknown feature %q\n", sanitizedFeature)
				os.Exit(1)
			}
		}
	}
```
3. Keep both `NewShared` and `New` functions:
```
// NewShared creates a new instance of the Testing Wrapper that provides additional
// functionality beyond the standard testing package.
// It doesn't require the testFeature param.
func NewShared(t NTB) *Wrapper {
	t.Helper()

	tw := &Wrapper{
		t: t,
	}
	return tw
}

// New creates a new instance of the Testing Wrapper that provides additional
// functionality beyond the standard testing package.
func New(t NTB, testFeature Feature) *Wrapper {
	t.Helper()

	if !KnownFeature(testFeature) {
		t.Fatalf("Test failed because the feature %q is unknown", testFeature)
	}

	skip := true
	for _, f := range TestFeatureArray {
		if f == testFeature {
			skip = false
		}
	}
	if skip {
		t.Skipf("Skip the test because the feature %q is not included in the test-features flag %q", testFeature, *e2e.TestFeatures)
	}

	tw := &Wrapper{
		t: t,
	}

	return tw
}
```

From simplicity's perspective, I couldn't see too much gain, but the check for allowed features spreads in multiple places.